state-sync/src/coordinator.rs

```diff
@@ -1340,84 +1332,234 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-/// is used locally for retrieving missing chunks for the local storage.
-#[derive(Clone)]
-pub struct SyncState {
-    committed_ledger_info: LedgerInfoWithSignatures,
-    synced_trees: ExecutedTrees,
-    trusted_epoch_state: EpochState,
-}
+#[cfg(test)]
+mod tests {
+    use crate::{client::SyncRequest, error::Error, shared_components::test_utils};
+    use anyhow::Result;
+    use diem_crypto::{
+        ed25519::{Ed25519PrivateKey, Ed25519Signature},
+        HashValue, PrivateKey, Uniform,
+    };
+    use diem_mempool::CommitResponse;
+    use diem_types::{
+        account_address::AccountAddress,
+        block_info::BlockInfo,
+        chain_id::ChainId,
+        ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
+        transaction::{
+            RawTransaction, Script, SignedTransaction, Transaction, TransactionPayload, Version,
+        },
+        waypoint::Waypoint,
+    };
+    use futures::{channel::oneshot, executor::block_on};
+    use std::{collections::BTreeMap, time::SystemTime};
+
+    #[test]
+    fn test_process_sync_request() {
+        let mut coordinator = test_utils::create_state_sync_coordinator_for_tests();
+
+        // Perform sync request for version that matches initial waypoint version
+        let (sync_request, mut callback_receiver) = create_sync_request_at_version(0);
+        coordinator.process_sync_request(sync_request).unwrap();
+        match callback_receiver.try_recv() {
+            Ok(Some(result)) => {
+                assert!(result.is_ok())
+            }
+            result => panic!("Expected okay but got: {:?}", result),
+        };
 
-impl SyncState {
-    pub fn new(
-        committed_ledger_info: LedgerInfoWithSignatures,
-        synced_trees: ExecutedTrees,
-        current_epoch_state: EpochState,
-    ) -> Self {
-        let trusted_epoch_state = committed_ledger_info
-            .ledger_info()
-            .next_epoch_state()
-            .cloned()
-            .unwrap_or(current_epoch_state);
-
-        SyncState {
-            committed_ledger_info,
-            synced_trees,
-            trusted_epoch_state,
-        }
+        // Set coordinator waypoint to version higher than storage
+        let waypoint_version = 10;
+        let waypoint_ledger_info = create_ledger_info_at_version(waypoint_version);
+        coordinator.waypoint = Waypoint::new_any(&waypoint_ledger_info.ledger_info());
+
+        // Verify coordinator won't process sync requests as it's not yet initialized
+        let (sync_request, mut callback_receiver) = create_sync_request_at_version(10);
+        match coordinator.process_sync_request(sync_request) {
+            Err(Error::UninitializedError(..)) => { /* Expected */ }
+            result => panic!("Expected an uninitialized error, but got: {:?}", result),
+        };
+        match callback_receiver.try_recv() {
+            Err(_) => { /* Expected */ }
+            result => panic!("Expected error but got: {:?}", result),
+        };
```

Comment:
Hey, something I learned about recently to make these cleaner, but not less lines (for future):

The `matches!` macro can let you do something like:
```
let result = callback_receiver.try_recv()
if !matches!(result, Err(_)) {
    panic!(...)
}
```
https://doc.rust-lang.org/std/macro.matches.html

Reply:
Oh nice! Looks like this landed in stable.. I saw it a while ago, but it was only in nightly 😄